### ISSUE TITLE
Some output of terser might be parsed wrongly by some tools

### ISSUE THREAD

**Issue by wkordalski** [terser/terser#1167]
**Bug report**

**Version**: try.terser.org

**Complete CLI command or `minify()` options used**

```js
{
  module: true,
  compress: {},
  mangle: {},
  output: {},
  parse: {},
  rename: {},
}
```

**`terser` input**

```js
function foo(cond) {
  if(cond) {
    return r=>({x:r.f});
  } else {
    return r=>({x:r.g});
  }
}
console.log(foo(confirm("Yes or No?"))({f: 7, g: 42}))
```

**`terser` output**

```js
console.log((confirm("Yes or No?")?o=>({x:o.f}):o=>({x:o.g}))({f:7,g:42}));
```

**Expected result**

Generally terser output looks good and Firefox executes such code correctly.
But some tools parse the code wrongly:
```js
console.log((
    confirm("Yes or No?")?
    o=> (
        ({x:o.f}):o => ({x:o.g})
    )
) ({f:7,g:42}) );
```

Editors (like VSCode) indicate errors:
 * on dot in expression `o.f`: `',' expected.`
 * on `o` in second expression `o=>`: `Type annotations can only be used in TypeScript files.`
 * on last closing parenthesis in expression `(confirm("Yes or No?")?o=>({x:o.f}):o=>({x:o.g}))`: `':' expected.`

How to reproduce: paste terser output into VSCode.

Webpack with some plugins inside:

> **Note:** Here I show output from build of a larger codebase as I don't have time to setup whole environment for test purposes.
> In the codebase terser generates an expression like the one presented above.

```
ERROR in /.../monaco-3640001003124600/js-prod/editor.js
./.../monaco-3640001003124600/js-prod/editor.js 1:1622053-1622054
[tsl] ERROR in /.../monaco-3640001003124600/js-prod/editor.js(1,1622054)
      TS8010: Type annotations can only be used in TypeScript files.
 @ ...

ERROR in /.../monaco-3640001003124600/js-prod/editor.js
./.../monaco-3640001003124600/js-prod/editor.js 1:1622011-1622012
[tsl] ERROR in /.../monaco-3640001003124600/js-prod/editor.js(1,1622012)
      TS1005: ',' expected.
 @ ...

ERROR in /.../monaco-3640001003124600/js-prod/editor.js
./.../monaco-3640001003124600/js-prod/editor.js 1:1622028-1622029
[tsl] ERROR in /.../monaco-3640001003124600/js-prod/editor.js(1,1622029)
      TS1005: ',' expected.
 @ ...

ERROR in /.../monaco-3640001003124600/js-prod/editor.js
./.../monaco-3640001003124600/js-prod/editor.js 1:1622043-1622044
[tsl] ERROR in /.../monaco-3640001003124600/js-prod/editor.js(1,1622044)
      TS1005: ',' expected.
 @ ...

ERROR in /.../monaco-3640001003124600/js-prod/editor.js
./.../monaco-3640001003124600/js-prod/editor.js 1:1622113-1622114
[tsl] ERROR in /.../monaco-3640001003124600/js-prod/editor.js(1,1622114)
      TS1005: ':' expected.
 @ ...
```

**Workaround**

It is enough to add parentheses to make everything fine:

```js
console.log((confirm("Yes or No?")?(o=>({x:o.f})):o=>({x:o.g}))({f:7,g:42}));
```

**Comment by jridgewell** [comment]
TS syntax is invalid with certain JS syntax, and unfortunately this isn't fixable. Terser is outputting JS, not TS, and some of the output will not be valid when parsed by a TS tool.

**Comment by jakebailey** [comment]
FYI, I reverted that fix, but for TS 4.8 one of the goals for fixing this will be to ensure that we prefer parsing what JS would parse had JS run on the code (rather than parsing so eagerly as return type annotations).
